### PR TITLE
139550-list-total-processed-payment-requests

### PR DESCRIPTION
**Pull request by simonnineteen90**
https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/139550

### CODE REVIEW

File: test/unit/data/events/scheme/get-submitted-events.test.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    const result = await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    expect(result[0].data).toEqual(extractedEvent.data)
+    expect(result[1].data).toEqual(enrichedEvent.data)
+  })
+
+  test('should return an empty array if no events', async () => {
+    mockListEntities.mockReturnValue([])
+    const result = await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    expect(result.length).toBe(0)
+  })
```
Comment (simonnineteen90) [inline test/unit/data/events/scheme/get-submitted-events.test.js:80]: Is there a way to test the mock table storage as explain in comments here?
Can you explain what exactly the imported `0data` /`mock0Data` is.

File: test/unit/data/events/scheme/get-submitted-events.test.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  // can this be tested because we are mocking the response from table storage?
+  test('should only return events that are of type submitted', async () => {
+    const result = await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    result.forEach(event => {
+      expect(event.type).toBe(PAYMENT_SUBMITTED)
+    })
+  })
```
Comment (johnwatson484) [inline test/unit/data/events/scheme/get-submitted-events.test.js:88]: Not with mocks in unit tests.  These tests are more testing the interface for table storage is being used as expected with the correct query parameters and the responses are handled appropriately. 

The type of test you're describing would be a local integration test in the `integration/local` folder.   You could load some events to storage, then use the SDK without mocks to see if only the submitted events are returned.

File: test/unit/data/events/scheme/get-submitted-events.test.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  // Instead can we only test what search parms the query is called with?
+  // test that it is called with the correct type
+  test('should call mockListEntities with type of submitted ', async () => {
+    await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    // what is the mock0data doing and where does it come from
+    expect(mockTableClient.listEntities).toHaveBeenCalledWith({ queryOptions: { filter: mockOdata`THIS CAN BE ANYTHING AND IT WONT FAIL${PAYMENT_SUBMITTED}` } })
```
Comment (johnwatson484) [inline test/unit/data/events/scheme/get-submitted-events.test.js:95]: For more detail on odata you can read through this -> https://www.odata.org/ but as a summary it's a protocol for how to query data.

Table storage uses it and they SDK contains a handy helper function to parse js into it.

Hadn't appreciated the awkwardness of testing with it for this scenario.  I'd say make this test just confirm that the query is called correctly and create a local integration test to cover the actual usage.

I'm not sure why there aren't any local integration tests already.  Suspect an oversight.

File: test/unit/data/events/scheme/get-events-by-scheme-id.test.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+const { SCHEME_ID: SCHEME_ID_CATEGORY } = require('../../../../../app/constants/categories')
+const submitted = require('../../../../mocks/events/submitted')
+const groupedEvent = require('../../../../mocks/events/grouped-event')
+const totalSchemeValues = require('../../../../mocks/events/total-scheme-values')
+
+const { getEventsBySchemeId } = require('../../../../../app/data/events/scheme-id/get-events-by-scheme-id')
+
+describe('get events by frn', () => {
+  beforeEach(() => {
+    jest.clearAllMocks()
```
Comment (simonnineteen90) [inline test/unit/data/events/scheme/get-events-by-scheme-id.test.js:26]: removing this `jest.clearAllMocks()` allows all tests to pass. Currently they all fail. Where is the correct place to rest the mocks or do i not need to in this scenario?

Comment (simonnineteen90) [reply test/unit/data/events/scheme/get-events-by-scheme-id.test.js:26]: needed an `await` on each test

File: app/data/events/scheme-id/get-submitted-events.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const { getClient, odata } = require('../../../storage')
+const { PAYMENT_EVENT } = require('../../../constants/event-types')
+const { PAYMENT_SUBMITTED } = require('../../../constants/events')
+
+const getSubmittedEvents = async (category) => {
+  const client = getClient(PAYMENT_EVENT)
+  const eventResults = client.listEntities({ queryOptions: { filter: odata`category eq '${category}' and type eq '${PAYMENT_SUBMITTED}'` } })
```
Comment (Kirsten1822) [inline app/data/events/scheme-id/get-submitted-events.js:7]: Could you explain what this line of code does?

File: test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
+const { FRN } = require('../../../../../mocks/values/frn')
+const { INVOICE_NUMBER } = require('../../../../../mocks/values/invoice-number')
+const { BPS, CS, SFI } = require('../../../../../../app/constants/schemes')
```
Comment (Kirsten1822) [inline test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js:3]: I'd put a space here to separate the mocks from the constants

File: test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+const schemeNames = require('../../../../../../app/constants/scheme-names')
+
+const { PAYMENT_EVENT, HOLD_EVENT, BATCH_EVENT, WARNING_EVENT } = require('../../../../../../app/constants/event-types')
```
Comment (Kirsten1822) [inline test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js:6]: These should probably be grouped with the other constants

File: test/mocks/events/total-scheme-values.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const { SCHEME_ID } = require('../values/scheme-id')
+const { VALUE } = require('../values/value')
+module.exports = {
```
Comment (Kirsten1822) [inline test/mocks/events/total-scheme-values.js:3]: Space needed here

File: test/unit/data/events/scheme/get-events-by-scheme-id.test.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+jest.mock('../../../../../app/data/events/scheme-id/get-total-scheme-values')
+const { getTotalSchemeValues: mockGetTotalSchemeValues } = require('../../../../../app/data/events/scheme-id/get-total-scheme-values')
+
+jest.mock('../../../../../app/data/events/scheme-id/sanitise-scheme-data')
+const { sanitiseSchemeData: mockSanitiseSchemeData } = require('../../../../../app/data/events/scheme-id/sanitise-scheme-data')
+
+const { SCHEME_ID: SCHEME_ID_CATEGORY } = require('../../../../../app/constants/categories')
+const submitted = require('../../../../mocks/events/submitted')
+const groupedEvent = require('../../../../mocks/events/grouped-event')
+const totalSchemeValues = require('../../../../mocks/events/total-scheme-values')
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/get-events-by-scheme-id.test.js:19]: I'd put these mocks above the constant, with a space to separate them

File: test/unit/data/events/scheme/get-events-by-scheme-id.test.js (hunk 4 above)
Comment (Kirsten1822) [reply test/unit/data/events/scheme/get-events-by-scheme-id.test.js:25]: So with the clearAllMocks the tests fail?

File: test/unit/data/events/scheme/get-submitted-events.test.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+const { PAYMENT_SUBMITTED } = require('../../../../../app/constants/events')
+const { PAYMENT_EVENT } = require('../../../../../app/constants/event-types')
+
+const { stringifyEventData } = require('../../../../helpers/stringify-event-data')
+
+const { getSubmittedEvents } = require('../../../../../app/data/events/scheme-id/get-submitted-events')
+
+const category = 'schemeId'
+let extractedEvent
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/get-submitted-events.test.js:20]: Space needed here

File: test/unit/data/events/scheme/get-total-scheme-values.test.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
+const { getTotalSchemeValues } = require('../../../../../app/data/events/scheme-id/get-total-scheme-values')
+const { BPS, CS } = require('../../../../../app/constants/schemes')
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/get-total-scheme-values.test.js:2]: Constant should be above getTotalSchemeValue - with a space to separate them

File: test/unit/data/events/scheme/get-submitted-events.test.js (hunk 2 above)
Comment (simonnineteen90) [reply test/unit/data/events/scheme/get-submitted-events.test.js]: Integration test added.

File: test/unit/data/events/scheme/get-events-by-scheme-id.test.js (hunk 4 above)
Comment (simonnineteen90) [reply test/unit/data/events/scheme/get-events-by-scheme-id.test.js:25]: no, they pass now i had missed off `await` when calling the function `getEventsBySchemeId` in the individual tests. I make this mistake roughly once a month 😅

File: test/unit/data/events/scheme/get-total-scheme-values.test.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    const result = getTotalSchemeValues(groupedEvents)
+    expect(result[0].paymentRequests).toBe(groupedEvents[0].events.length)
+  })
+
+  test('schemeId should be equal to groupedEvents schemeId ', () => {
+    const result = getTotalSchemeValues(groupedEvents)
+    expect(result[0].schemeId).toBe(groupedEvents[0].schemeId)
+  })
+
+  test('value should be equal to sum of events values when there are 3 events', () => {
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/get-total-scheme-values.test.js:57]: Why only 3 events?

File: test/unit/data/events/scheme/group-events-by-scheme.test.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
+const { BPS, CS, SFI } = require('../../../../../app/constants/schemes')
+const submitted = require('../../../../mocks/events/submitted')
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/group-events-by-scheme.test.js:2]: I'd put the mock at the top of the file with a space between to separate it from the constant

File: test/unit/data/events/scheme/group-events-by-scheme.test.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    })
+
+    events = [submitted, submitted]
+    bpsEvents = createEventsForScheme(BPS, events)
+    csEvents = createEventsForScheme(CS, events)
+    sfiEvents = createEventsForScheme(SFI, events)
+    mixedSchemeEvents = [...bpsEvents, ...csEvents, ...sfiEvents]
+  })
+
+  test('returned schemeId should be equal to partitionKey of all events in eventGroup', () => {
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/group-events-by-scheme.test.js:28]: Would 'should return schemeId equal to partitionKey of all events in eventGroup' work?

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
+
+const { orderGroupedEventsByScheme } = require('../../../../../app/data/events/scheme-id/order-grouped-events-by-scheme')
+const { SFI, SFI_PILOT, LUMP_SUMS, VET_VISITS, CS, BPS, FDMR } = require('../../../../../app/constants/schemes')
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:3]: I'd put the constant at the top of the file and separate it with a space

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+let groupedEvents
+let groupedEvent
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:6]: I'd swap the order of these two around, since groupedEvents appears last below

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    sfiPilotEvent = { ...groupedEvent, schemeId: SFI_PILOT }
+    lumpSumsEvent = { ...groupedEvent, schemeId: LUMP_SUMS }
+    vetEvent = { ...groupedEvent, schemeId: VET_VISITS }
+    csEvent = { ...groupedEvent, schemeId: CS }
+    bpsEvent = { ...groupedEvent, schemeId: BPS }
+    fdmrEvent = { ...groupedEvent, schemeId: FDMR }
+    groupedEvents = [bpsEvent, vetEvent, sfiEvent, sfiPilotEvent, lumpSumsEvent, csEvent, fdmrEvent]
+  })
+
+  test('should order grouped events into sequence by schemeId when all schemes present', () => {
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:29]: What does 'by sequence' mean here?

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    expect(orderedGroupedEvents[0].schemeId).toEqual(SFI)
+    expect(orderedGroupedEvents[1].schemeId).toEqual(SFI_PILOT)
+    expect(orderedGroupedEvents[2].schemeId).toEqual(LUMP_SUMS)
+    expect(orderedGroupedEvents[3].schemeId).toEqual(VET_VISITS)
+    expect(orderedGroupedEvents[4].schemeId).toEqual(CS)
+    expect(orderedGroupedEvents[5].schemeId).toEqual(BPS)
+    expect(orderedGroupedEvents[6].schemeId).toEqual(FDMR)
+  })
+
+  test('should order grouped events into sequence by schemeId when only two schemes present', () => {
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:40]: See my comment here: https://github.com/DEFRA/ffc-pay-data-hub/pull/2/files#r1212816728

File: test/unit/data/events/scheme/sanitise-scheme-data.test.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+const schemeNames = require('../../../../../app/constants/scheme-names')
+const { sanitiseSchemeData } = require('../../../../../app/data/events/scheme-id/sanitise-scheme-data')
+
+const totalSchemeValues = require('../../../../mocks/events/total-scheme-values')
+
+describe('get events by frn', () => {
+  beforeEach(() => {
+
+  })
```
Comment (Kirsten1822) [inline test/unit/data/events/scheme/sanitise-scheme-data.test.js:12]: Do you need this beforeEach if it's empty?

File: test/unit/data/get-data.test.js [hunk 20]
```diff
@@ -26,11 +28,16 @@ describe('get data', () => {
     expect(mockGetEventsByCorrelationId).toHaveBeenCalledWith(CORRELATION_ID_VALUE)
   })
 
-  test('should get batch events for batch category', async () => {
+  test('should get schemeId events for scheme category', async () => {
     await getData(BATCH_CATEGORY, BATCH_VALUE)
     expect(mockGetEventsByBatch).toHaveBeenCalledWith(BATCH_VALUE)
   })
```
Comment (Kirsten1822) [inline test/unit/data/get-data.test.js:34]: I'm not sure why this one was renamed here

File: app/data/events/scheme-id/get-submitted-events.js (hunk 5 above)
Comment (simonnineteen90) [reply app/data/events/scheme-id/get-submitted-events.js:7]: `client` is defined on line 6 and is getting the payment instance of the Azure Table Storage, this `client` has different functions we can call to interact with the data that is stored in the table.

line 7 is using the `listEntities` function to get all the 'entities' (which in our case are the events). We are then passing in an object which contains query data to filter the events. Specifically we are telling it to filter and only return the events that have a category of category (`schemeId`) and type of payment_submitted (`uk.gov.defra.ffc.pay.payment.submitted`)

So basically we use this line to get the events from the table storage based on the filters we pass in.

File: test/unit/data/events/scheme/get-events-by-scheme-id.test.js (hunk 4 above)
Comment (Kirsten1822) [reply test/unit/data/events/scheme/get-events-by-scheme-id.test.js:25]: You're fine!

File: app/data/events/scheme-id/get-submitted-events.js (hunk 5 above)
Comment (Kirsten1822) [reply app/data/events/scheme-id/get-submitted-events.js:7]: Thanks!

File: test/mocks/events/total-scheme-values.js (hunk 8 above)
Comment (simonnineteen90) [reply test/mocks/events/total-scheme-values.js:3]: also moved this out of the `events` folder as its not an event, doesnt really fit the other folders so put it in the root `mocks` folder

File: test/unit/data/events/scheme/get-total-scheme-values.test.js [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+    const result = getTotalSchemeValues(groupedEvents)
+    expect(result[0].value).toBe(300000)
+  })
+
+  test('value should be equal to sum of events values when there is only one event', () => {
```
[the remark this answers is not in the document]
Comment (simonnineteen90) [reply test/unit/data/events/scheme/get-total-scheme-values.test.js:62]: there are multiple events for bps but i have removed two of the events with `.pop` for the first element (cs) of `groupedEvents` which is the part I am testing.

File: test/unit/data/events/scheme/group-events-by-scheme.test.js (hunk 14 above)
Comment (simonnineteen90) [reply test/unit/data/events/scheme/group-events-by-scheme.test.js:28]: yep

**Comment by simonnineteen90** [comment]
have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo

File: app/data/events/scheme-id/get-total-scheme-values.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
+const getTotalSchemeValues = (groupedEvents) => {
+  return groupedEvents.map((eventsByScheme) => {
+    const value = eventsByScheme.events.reduce((total, current) => {
```
[the remark this answers is not in the document]
Comment (Kirsten1822) [reply app/data/events/scheme-id/get-total-scheme-values.js:3]: Thanks :)

File: test/unit/data/events/scheme/get-total-scheme-values.test.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (Kirsten1822) [reply test/unit/data/events/scheme/get-total-scheme-values.test.js:64]: Then maybe it would be better to rename this test to '...only one event in first element of groupedEvents'?

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js (hunk 17 above)
Comment (Kirsten1822) [reply test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:29]: Would schemeId be in ascending order then?

Comment (simonnineteen90) [reply test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:29]: yep, have added that to test name

File: test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js (hunk 18 above)
Comment (simonnineteen90) [reply test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:40]: updated to include ascending in description

**Comment by Kirsten1822** [comment]
> have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo

Will you need to change the test file names also?

**Comment by simonnineteen90** [comment]
> > have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo
> 
> Will you need to change the test file names also?

So some references to `schemeId` in the test names are correct as we are still using the `schemeId` for some tests but have been through and changed any that should just be `scheme`.

**Review by Kirsten1822 — APPROVED** [review]
Happy to approve :)
